conans/client/build/compiler_id.py

```diff
@@ -0,0 +1,233 @@
+        return CompilerId(compiler, major, minor, patch)
+    except KeyError:
+        return UNKNOWN_COMPILER
+    except ValueError:
+        return UNKNOWN_COMPILER
+    except TypeError:
+        return UNKNOWN_COMPILER
+
+
+def _detect_compiler_id_internal(executable, runner=None):
+    runner = runner or ConanRunner()
+    # use a temporary file, as /dev/null might not be available on all platforms
+    tmpname = tempfile.mktemp(suffix=".c")
+    with open(tmpname, "wb") as f:
+        f.write(b"\n")
+
+    cmd = tempfile.mktemp(suffix=".cmd")
+    with open(cmd, "wb") as f:
+        f.write(b"echo off\nset MSC_CMD_FLAGS\n")
+
+    detectors = [
+        # "-dM" generate list of #define directives
+        # "-E" run only preprocessor
+        # "-x c" compiler as C code
+        # the output is of lines in form of "#define name value"
+        "-dM -E -x c",
+        "--driver-mode=g++ -dM -E -x c",  # clang-cl
+        "-c -xdumpmacros",  # SunCC,
+        # cl (Visual Studio, MSVC)
+        # "/nologo" Suppress Startup Banner
+        # "/E" Preprocess to stdout
+        # "/B1" C front-end
+        # "/c" Compile Without Linking
+        # "/TC" Specify Source File Type
+        '/nologo /E /B1 "%s" /c /TC' % cmd,
+        "/QdM /E /TC"  # icc (Intel) on Windows,
+        "-Wp,-dM -E -x c"  # QNX QCC
+    ]
+    try:
+        for detector in detectors:
+            command = '%s %s "%s"' % (executable, detector, tmpname)
+            result = StringIO()
+            if 0 == runner(command, output=result):
+                output = result.getvalue()
+                defines = dict()
+                for line in output.splitlines():
+                    tokens = line.split(' ', 3)
+                    if len(tokens) == 3 and tokens[0] == '#define':
+                        defines[tokens[1]] = tokens[2]
+                    # MSVC dumps macro definitions in single line:
+                    # "MSC_CMD_FLAGS=-D_MSC_VER=1921 -Ze"
+                    elif line.startswith("MSC_CMD_FLAGS="):
+                        line = line[len("MSC_CMD_FLAGS="):].rstrip()
+                        defines = dict()
+                        tokens = line.split()
+                        for token in tokens:
+                            if token.startswith("-D") or token.startswith("/D"):
+                                token = token[2:]
+                                if '=' in token:
+                                    name, value = token.split('=', 2)
+                                else:
+                                    name, value = token, '1'
+                                defines[name] = value
+                        break
+                compiler = _parse_compiler_version(defines)
+                if compiler == UNKNOWN_COMPILER:
+                    continue
+                return compiler
+        return UNKNOWN_COMPILER
+    finally:
+        os.unlink(tmpname)
+        os.unlink(cmd)
+
+
+_compiler_id_cache = dict()
```

Comment:
so the caching will only work for the detection of the default profile, right? I mean, the function `detect_compiler_id()` will be called just once, so the caching will not make much difference as the results are not preserved for the next Conan execution

Reply:
indeed. the main reason I have added this is that in tests default profile was initialized thousands of times within the same process (pytest). so for testing purposes only it makes sense to have a caching here.
for the real invocation (such as `conan profile new <xxx> --detect`) it will be always new process and cache always be empty.
this is something I would expect, as for instance if you did some system update, you may have `gcc` pointing to newer version 11, while previously it was pointing to 10. so in this case detection will work as expected.